# 
# SVN commit 1032802 by sengels:
# 
# fix missing rename
# 

Index: knotes/knoteslegacy.cpp
===================================================================
--- knotes/knoteslegacy.cpp	(révision 1032801)
+++ knotes/knoteslegacy.cpp	(révision 1032802)
@@ -243,7 +243,7 @@
   // update the config
   KConfig config( configFile );
   KConfigGroup cfggrp( &config, "Data" );
-  journal->setSummary( grp.readEntry( "name" ) );
+  journal->setSummary( cfggrp.readEntry( "name" ) );
   config.deleteGroup( "Data", KConfig::Localized );
   KConfigGroup cfggrp2(&config, "General" ); // XXX right?
   cfggrp2.writeEntry( "version", KNOTES_VERSION );
