# 
# SVN commit 1032446 by sengels:
# 
# templates can't be exported
# 

Index: libkdepim/pluginloader.h
===================================================================
--- libkdepim/pluginloader.h	(révision 1032445)
+++ libkdepim/pluginloader.h	(révision 1032446)
@@ -77,7 +77,7 @@
    *
    **/
   template< typename T, typename T_config >
-  class KDE_EXPORT PluginLoader : public PluginLoaderBase {
+  class PluginLoader : public PluginLoaderBase {
   protected:
     PluginLoader() : PluginLoaderBase() {}
 
